File: Optical-Flow-in-OpenCV/algorithms/lucas-kanade.cpp
// Sparse pyramidal Lucas-Kanade tracking for the optical-flow tutorial.

#include "optical_flow.hpp"

#include <opencv2/core.hpp>
#include <opencv2/core/version.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#if CV_VERSION_MAJOR >= 5
#include <opencv2/features.hpp>
#endif
#include <opencv2/video.hpp>
#include <opencv2/videoio.hpp>

#include <cmath>
#include <filesystem>
#include <numeric>
#include <stdexcept>
#include <vector>

namespace optical_flow {

namespace {

namespace fs = std::filesystem;

void validate_summary(const Summary& summary) {
  if (summary.frame_pairs <= 0) {
    throw std::runtime_error("No feature tracks were produced from the video");
  }
  if (!std::isfinite(summary.mean_magnitude) ||
      summary.mean_magnitude <= 0.0) {
    throw std::runtime_error(
        "Expected finite nonzero tracked motion, got " +
        std::to_string(summary.mean_magnitude));
  }
  const cv::Mat reloaded =
      cv::imread(summary.output_path.string(), cv::IMREAD_COLOR);
  if (reloaded.empty()) {
    throw std::runtime_error("Saved sparse-flow visualization is unreadable");
  }
}

}  // namespace

Summary run_lucas_kanade(const RunOptions& options) {
  cv::VideoCapture capture(options.video_path.string());
  if (!capture.isOpened()) {
    throw std::runtime_error(
        "Unable to open input video: " + options.video_path.string());
  }

  cv::Mat previous_bgr;
  if (!capture.read(previous_bgr) || previous_bgr.empty()) {
    throw std::runtime_error(
        "Input video has no readable frames: " + options.video_path.string());
  }

  cv::Mat previous_gray;
  cv::cvtColor(previous_bgr, previous_gray, cv::COLOR_BGR2GRAY);
  std::vector<cv::Point2f> previous_points;
  cv::goodFeaturesToTrack(
      previous_gray,
      previous_points,
      100,
      0.3,
      7.0,
      cv::noArray(),
      7);
  if (previous_points.empty()) {
    throw std::runtime_error(
        "No Shi-Tomasi features were found in the first frame");
  }

  cv::RNG generator(7);
  std::vector<cv::Scalar> colors;
  colors.reserve(100);
  for (int index = 0; index < 100; ++index) {
    colors.emplace_back(
        generator.uniform(0, 256),
        generator.uniform(0, 256),
        generator.uniform(0, 256));
  }

  cv::Mat trails = cv::Mat::zeros(previous_bgr.size(), previous_bgr.type());
  cv::Mat visualization = previous_bgr.clone();
  std::vector<double> magnitudes;
  int frame_pairs = 0;

  while (options.max_frames == 0 || frame_pairs < options.max_frames) {
    cv::Mat current_bgr;
    if (!capture.read(current_bgr) || current_bgr.empty()) {
      break;
    }

    cv::Mat current_gray;
    cv::cvtColor(current_bgr, current_gray, cv::COLOR_BGR2GRAY);
    std::vector<cv::Point2f> current_points;
    std::vector<unsigned char> status;
    std::vector<float> errors;
    cv::calcOpticalFlowPyrLK(
        previous_gray,
        current_gray,
        previous_points,
        current_points,
        status,
        errors,
        cv::Size(15, 15),
        2,
        cv::TermCriteria(
            cv::TermCriteria::COUNT | cv::TermCriteria::EPS,
            10,
            0.03));

    std::vector<cv::Point2f> good_new;
    std::vector<double> frame_magnitudes;
    cv::Mat frame = current_bgr.clone();
    for (std::size_t index = 0; index < status.size(); ++index) {
      if (status[index] == 0U) {
        continue;
      }
      good_new.push_back(current_points[index]);
      const cv::Point2f displacement =
          current_points[index] - previous_points[index];
      frame_magnitudes.push_back(cv::norm(displacement));
      const cv::Scalar& color = colors[index % colors.size()];
      cv::line(
          trails,
          current_points[index],
          previous_points[index],
          color,
          2,
          cv::LINE_AA);
      cv::circle(
          frame,
          current_points[index],
          4,
          color,
          cv::FILLED,
          cv::LINE_AA);
    }
    if (good_new.empty()) {
      break;
    }

    const double frame_mean =
        std::accumulate(
            frame_magnitudes.begin(),
            frame_magnitudes.end(),
            0.0) /
        static_cast<double>(frame_magnitudes.size());
    magnitudes.push_back(frame_mean);
    cv::add(frame, trails, visualization);
    ++frame_pairs;

    if (options.show_windows) {
      cv::imshow("Sparse Lucas-Kanade optical flow", visualization);
      const int key = cv::waitKey(25);
      if (key == 27 || key == 'q') {
        break;
      }
      if (key == 'c') {
        trails.setTo(cv::Scalar::all(0));
      }
    }

    previous_gray = current_gray;
    previous_points = std::move(good_new);
  }

  capture.release();
  if (options.show_windows) {
    cv::destroyAllWindows();
  }
  if (frame_pairs == 0 || magnitudes.empty()) {
    throw std::runtime_error("No feature tracks were produced from the video");
  }

  fs::create_directories(options.output_dir);
  const fs::path output_path =
      options.output_dir / "lucaskanade-optical-flow.png";
  if (!cv::imwrite(output_path.string(), visualization)) {
    throw std::runtime_error(
        "Unable to write optical-flow image: " + output_path.string());
  }

  const double mean_magnitude =
      std::accumulate(magnitudes.begin(), magnitudes.end(), 0.0) /
      static_cast<double>(magnitudes.size());
  const Summary summary{frame_pairs, mean_magnitude, output_path};
  if (options.validate) {
    validate_summary(summary);
  }
  return summary;
}

}  // namespace optical_flow

File: Optical-Flow-in-OpenCV/algorithms/lucas_kanade.py
"""Sparse pyramidal Lucas-Kanade tracking for the optical-flow tutorial."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
import math

import cv2
import numpy as np


@dataclass(frozen=True)
class SparseFlowSummary:
    """Stable metrics returned to the CLI and regression tests."""

    frame_pairs: int
    mean_magnitude: float
    output_path: Path


def run_lucas_kanade(
    *,
    video_path: Path,
    output_dir: Path,
    show_windows: bool,
    max_frames: int,
    validate: bool,
) -> SparseFlowSummary:
    """Track good features across a video and save the final trail image."""

    capture = cv2.VideoCapture(str(video_path))
    if not capture.isOpened():
        raise FileNotFoundError(f"Unable to open input video: {video_path}")

    has_frame, previous_bgr = capture.read()
    if not has_frame or previous_bgr is None:
        capture.release()
        raise RuntimeError(f"Input video has no readable frames: {video_path}")

    previous_gray = cv2.cvtColor(previous_bgr, cv2.COLOR_BGR2GRAY)
    previous_points = cv2.goodFeaturesToTrack(
        previous_gray,
        maxCorners=100,
        qualityLevel=0.3,
        minDistance=7,
        blockSize=7,
    )
    if previous_points is None:
        capture.release()
        raise RuntimeError("No Shi-Tomasi features were found in the first frame")

    generator = np.random.default_rng(7)
    colors = generator.integers(0, 256, size=(100, 3), dtype=np.uint8)
    trails = np.zeros_like(previous_bgr)
    frame_pairs = 0
    magnitudes: list[float] = []
    visualization = previous_bgr.copy()

    while max_frames == 0 or frame_pairs < max_frames:
        has_frame, current_bgr = capture.read()
        if not has_frame or current_bgr is None:
            break

        current_gray = cv2.cvtColor(current_bgr, cv2.COLOR_BGR2GRAY)
        current_points, status, _ = cv2.calcOpticalFlowPyrLK(
            previous_gray,
            current_gray,
            previous_points,
            None,
            winSize=(15, 15),
            maxLevel=2,
            criteria=(
                cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT,
                10,
                0.03,
            ),
        )
        if current_points is None or status is None:
            break

        valid = status.reshape(-1) == 1
        good_new = current_points[valid]
        good_old = previous_points[valid]
        if len(good_new) == 0:
            break

        displacement = good_new.reshape(-1, 2) - good_old.reshape(-1, 2)
        magnitudes.append(float(np.mean(np.linalg.norm(displacement, axis=1))))

        frame = current_bgr.copy()
        for index, (new, old) in enumerate(zip(good_new, good_old)):
            new_xy = tuple(np.rint(new.ravel()).astype(int))
            old_xy = tuple(np.rint(old.ravel()).astype(int))
            color = colors[index % len(colors)].tolist()
            cv2.line(trails, new_xy, old_xy, color, 2, cv2.LINE_AA)
            cv2.circle(frame, new_xy, 4, color, -1, cv2.LINE_AA)

        visualization = cv2.add(frame, trails)
        frame_pairs += 1

        if show_windows:
            cv2.imshow("Sparse Lucas-Kanade optical flow", visualization)
            key = cv2.waitKey(25) & 0xFF
            if key in (27, ord("q")):
                break
            if key == ord("c"):
                trails.fill(0)

        previous_gray = current_gray
        previous_points = good_new.reshape(-1, 1, 2)

    capture.release()
    if show_windows:
        cv2.destroyAllWindows()

    if frame_pairs == 0:
        raise RuntimeError("No feature tracks were produced from the video")

    output_dir.mkdir(parents=True, exist_ok=True)
    output_path = output_dir / "lucaskanade-optical-flow.png"
    if not cv2.imwrite(str(output_path), visualization):
        raise OSError(f"Unable to write optical-flow image: {output_path}")

    overall_mean = float(np.mean(magnitudes))
    if validate:
        if not math.isfinite(overall_mean) or overall_mean <= 0.0:
            raise RuntimeError(
                f"Expected finite nonzero tracked motion, got {overall_mean}"
            )
        reloaded = cv2.imread(str(output_path), cv2.IMREAD_COLOR)
        if reloaded is None or reloaded.size == 0:
            raise RuntimeError("Saved sparse-flow visualization is unreadable")

    return SparseFlowSummary(frame_pairs, overall_mean, output_path)
